src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,266 @@
+public:
+    using IRMutator::visit;
+
+    Stmt visit(const Atomic *op) override {
+        // Collect the names of all Store nodes inside.
+        CollectStoreNames collector;
+        op->body.accept(&collector);
+        // Search for let bindings that access the providers.
+        FindAtomicLetBindings finder(collector.store_names);
+        op->body.accept(&finder);
+        if (finder.found) {
+            // Can't remove mutex lock. Leave the Stmt as is.
+            return IRMutator::visit(op);
+        } else {
+            remove_mutex_lock_names.insert(op->mutex_name);
+            Stmt body = mutate(op->body);
+            return Atomic::make(op->producer_name,
+                                "",
+                                {},
+                                op->tuple_size,
+                                op->dimensions,
+                                std::move(body));
+        }
+    }
+
+    std::set<string> remove_mutex_lock_names;
+};
+
+class FindAtomicMutexUsage : public IRGraphVisitor {
+public:
+    FindAtomicMutexUsage(const std::string &producer_name)
+        : producer_name(producer_name) {}
+
+    using IRGraphVisitor::visit;
+
+    void visit(const Atomic *op) override {
+        for (Expr i : op->mutex_indices) {
+            include(i);
+        }
+        include(op->body);
+        if (op->producer_name == producer_name &&
+                op->mutex_name != "") {
+            if (found) {
+                // Multiple atomics inside the producer,
+                // make sure they have consistent information
+                internal_assert(mutex_name == op->mutex_name &&
+                    tuple_size == op->tuple_size &&
+                    dimensions == op->dimensions) <<
+                    "Inconsistent informatino of atomics inside a Producer node.\n";
+            }
+            found = true;
+            mutex_name = op->mutex_name;
+            tuple_size = op->tuple_size;
+            dimensions = op->dimensions;
+        }
+    }
+
+    const std::string &producer_name;
+    bool found = false;
+    std::string mutex_name;
+    int tuple_size = 0;
+    int dimensions = 0;
+};
+
+class AddAtomicMutex : public IRMutator {
+public:
+    using IRMutator::visit;
+
+    const Target &target;
+
+    AddAtomicMutex(const Target &target) : target(target) {}
+
+    Stmt visit(const ProducerConsumer *op) override {
+        FindAtomicMutexUsage finder(op->name);
+        if (op->is_producer) {
+            finder.visit(op);
+            if (!finder.found) {
+                return IRMutator::visit(op);
+            }
+        } else {
+            return IRMutator::visit(op);
+        }
+
+        const std::string &producer_name = finder.producer_name;
+        const std::string &mutex_name = finder.mutex_name;
+        int tuple_size = finder.tuple_size;
+        int dimensions = finder.dimensions;
+        const char *extent_field = tuple_size == 1 ? ".extent." : ".0.extent.";
```

Comment:
Constructing these by name is a red flag. I think if you 1) Do this at the allocate node and get the extent from the allocate node's extent, and 2) remove the indices from the atomic node in the IR and just lock whatever single-dimensional indices are stored to inside the body, then you won't need any of these.

Reply:
I don't see an easy way to get rid of this at the moment if the corresponding buffer is the output.

Reply:
Also I remember that the llvm codegen assumes all buffers have these fields when packing the closure. If I remove these it will likely trigger an assertion inside the codegen.

Reply:
I'll store the extent in the atomic node if the producer is the output.

Reply:
That doesn't work either because I'm just essentially moving these statements to where ever I construct that extent. In general Halide doesn't store the output extents in any IR nodes except these variables which you can only find by name.

Reply:
I'll pass the output funcs/output buffers inside the AddAtomicMutex pass.

Reply:
Done. See my comment below.